proof/refine/ARM/IpcCancel_R.thy

```diff
@@ -1954,22 +1438,23 @@ proof -
         apply (rule ep_cancel_corres_helper)
        apply (rule mapM_x_wp')
        apply (wp weak_sch_act_wf_lift_linear set_thread_state_runnable_weak_valid_sched_action | simp)+
-      apply (rule_tac R="\<lambda>_ s. \<forall>x\<in>set list. tcb_at' x s \<and> valid_objs' s"
+      apply (rule_tac R="\<lambda>_ s. \<forall>x\<in>set list. tcb_at' x s \<and> valid_objs' s \<and> pspace_aligned' s \<and> pspace_distinct' s"
                    in hoare_post_add)
       apply (rule mapM_x_wp')
-      apply (rule hoare_name_pre_state)
-      apply ((wp hoare_vcg_const_Ball_lift mapM_x_wp'
-                sts_valid_queues setThreadState_not_st sts_st_tcb' tcbSchedEnqueue_not_st
-           | clarsimp
-           | fastforce elim: obj_at'_weakenE simp: valid_tcb_state'_def)+)[2]
-    apply (rule hoare_name_pre_state)
+      apply ((wpsimp wp: hoare_vcg_const_Ball_lift mapM_x_wp' sts_st_tcb' sts_valid_objs'
+                   simp: valid_tcb_state'_def
+              | strengthen valid_objs'_valid_tcbs')+)[3]
+     apply fastforce
     apply (wp hoare_vcg_const_Ball_lift set_ep_valid_objs'
         | (clarsimp simp: valid_ep'_def)
-        | (drule (1) bspec, clarsimp simp: valid_pspace'_def valid_tcb'_def valid_ep'_def elim!: valid_objs_valid_tcbE))+
+        | (drule (1) bspec, clarsimp simp: valid_pspace'_def valid_tcb'_def valid_ep'_def
+        | strengthen valid_objs'_valid_tcbs'))+
```

Comment:
Feel free to ignore, but since you've already restyled some of this lemma you could do this apply step as well.